### ISSUE TITLE
Nohup error message

### ISSUE THREAD

**Issue by tunetheweb** [matomo-org/plugin-GoogleAnalyticsImporter#173]
Any ideas why we have this:

https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/7927ff84db063ce77c2d4b7daa004bf8f8801692/Diagnostic/RequiredExecutablesCheck.php#L63-L66

Since `which nohup` will return 0 when it's successul, and non-0 when it can't find the binary isn't this wrong?

Woudn't it be better as:

```php
    private function isNohupPresent()
    {
        return (bool) @shell_exec('which nohup') == 0;
    }
```

This lead to me having a critical error message for ages, even though the import was working:

> [GoogleAnalyticsImporter] Required Executables (nohup) Matomo cannot find the nohup executable. The Google Analytics Importer uses nohup to execute the import process so it will continue on its own after a web request finishes. This executable must be present and accessible from PHP in order for the importer to function.

I finally decided to tackle and I've manually applied that change and it's gone. Woo hoo!

Happy to submit a PR for this if you agree with the fix but notsure why no one else is complaining about this?

**Comment by Findus23** [comment]
I think (but I am not entirely sure as I am not that familiar with PHP) the confusion is that you think that `shell_exec` will return the return code of the command. But it returns the output of the command which is empty (converting to false) if it is missing and the path to the binary if it exists (which converts to true).

So the function works as it should (as long as the shell used by PHP has a `which` builtin).

You can run `php -r "var_dump(shell_exec('which nohup'));"` to verify this.

**Comment by tunetheweb** [comment]
Ah you are correct sorry.

So it appears my php cannot find `which`. As soon as I change it to this it works:

    private function isNohupPresent()
    {
        putenv('PATH=/usr/bin');
        return (bool) @shell_exec('which nohup');
    }

Looking into my setup so avoid having to add that.

Will close this anyway. Thanks for getting back so quickly!

**Comment by tunetheweb** [comment]
FYI I had to uncomment this line in /etc/php-fpm.d/www.conf and then bounce the php-fpm service

```
;env[PATH] = /usr/local/bin:/usr/bin:/bin
```

All working now with the original code.

So guess the error message was right (it couldn't find `nohup`), and I should have stopped making excuses and read it and fixed the problem it was complaining about! 🤣

**Comment by riczorn** [comment]
I have the same error message, so I created a php file in the root like:

`echo shell_exec('nohup php -v');`

and it correctly returns: `PHP 8.1.14 (cli) (built: Jan 4 2023 06:45:14) (NTS gcc x86_64) Copyright (c)...`

so I should guess nohup and php are available and in the path.

Matomo version : 4.13.3
Google Analytics importer version 4.5.0
PHP 8.1.14

**Comment by AltamashShaikh** [comment]
@riczorn we are using `shell_exec('which nohup')` does that work too ?

**Comment by riczorn** [comment]
well we want to ensure nohup and php are available to the webserver, so you can make separate calls to pinpoint issues, and echo so you don't have to check logs

echo shell_exec('nohup --version');

**Comment by riczorn** [comment]
well we want to ensure nohup and php are available to the webserver, so you can make separate calls to pinpoint issues, and echo so you don't have to check logs

`echo shell_exec('nohup --version');`

**Comment by AltamashShaikh** [comment]
@riczorn sorry I didn't get you

**Comment by riczorn** [comment]
@AltamashShaikh in your first comment, you mentioned you were invoking `which nohup`.
Google Importer requires nohup and php-cgi be available to the webserver. Hence testing for the existance of which brings no useful information. What you want to test is that php and nohup are available, which is what my command does in a single call: `nohup php -v` invokes php from nohup and `echo` displays the result on the page, so you can confirm in one call that everything is in place.
Alas, this changes nothing as despite confirming that everything is fine on my system, I still get the dreaded error. Still, the import seems to be running so it may just be something we can ignore.

**Comment by AltamashShaikh** [comment]
@riczorn understood, you can disable the use of nohup, see this [code](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/4.x-dev/config/config.php#L12) and we would like to keep this checks individually rather than combining them

**Comment by riczorn** [comment]
I believe that exposing an option for nohup is unnecessary. It could start with nohup disabled, then test for the shell function to use:

```
$shell = '';
if (function_exists('exec')) $shell = 'exec';
if (function_exists('shell_exec')) $shell = 'shell_exec';
```

and maybe create a function to invoke the right one.
Then `if (!empty($shell))`

```
$is_nohup_available = shell_exec('nohup --version');
$is_php_available = shell_exec('php -v');
```

will give us enough information, if both the latter tests return an acceptable version then nohup support can be enabled without asking the user.

**Comment by AltamashShaikh** [comment]
@riczorn can you send a PR with the above change ?
We can review and merge it

**Comment by riczorn** [comment]
I added the nohup detection, https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/pull/362
I looked around and it would be appropriate to add a similar test for php, but you are invoking the matomo core directly `$cliPhp->findPhpBinary()`. Since it seems this is the preferred way in matomo for longer tasks, I guess that change should go in `cliPhp`.

Ideally, since `nohup` and `php` are invoked together, I would add the `php` test and `nohup` tests all in the matomo core, and expose a function in cliPhp such as runPhp() where you just pass in the payload to avoid code duplication. But I don't know enough of the matomo architecture to determine whether this is feasible, also I don't understand when it uses `shell_exec`, `exec` and `passthrough`.
Without going to the core, things could still be made easier just by creating a `Task.php: execPhp($phpCommand)`  and let it do all the testing.
